### PR TITLE
expression: make greatest/least type comparison compatible with MySQL

### LINKED ISSUE #17994: Incompatible behaviour of least() and greatest() functions with MySQL
### 1. Minimal reproduce step (Required)

```sql
drop table t;
create table t (int_type int, str_type char(255), datetime_type datetime);
insert into t values (12345, "invalid_time", "2020-01-01");
select int_type, str_type, datetime_type from t;
select least(int_type, str_type, datetime_type) from t;
select least(int_type, str_type) from t;
select least(int_type, datetime_type) from t;
select least(str_type, datetime_type) from t;
select greatest(int_type, str_type, datetime_type) from t;
select greatest(int_type, str_type) from t;
select greatest(int_type, datetime_type) from t;
select greatest(str_type, datetime_type) from t;
```

I tried these queries on MySQL and TiDB respectively.

### 2. What did you expect to see? (Required)

mysql  Ver 8.0.19 for osx10.15 on x86_64 (Homebrew)

In MySQL:

```sql
+----------+--------------+---------------------+
| int_type | str_type     | datetime_type       |
+----------+--------------+---------------------+
|    12345 | invalid_time | 2020-01-01 00:00:00 |
+----------+--------------+---------------------+
1 row in set (0.00 sec)

+------------------------------------------+
| least(int_type, str_type, datetime_type) |
+------------------------------------------+
| 12345                                    |
+------------------------------------------+
1 row in set, 2 warnings (0.00 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
+---------------------------+
| least(int_type, str_type) |
+---------------------------+
| 12345                     |
+---------------------------+
1 row in set (0.00 sec)

+--------------------------------+
| least(int_type, datetime_type) |
+--------------------------------+
| 12345                          |
+--------------------------------+
1 row in set, 1 warning (0.01 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
+--------------------------------+
| least(str_type, datetime_type) |
+--------------------------------+
| 2020-01-01 00:00:00            |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
+---------------------------------------------+
| greatest(int_type, str_type, datetime_type) |
+---------------------------------------------+
| 2020-01-01 00:00:00.000000                  |
+---------------------------------------------+
1 row in set, 2 warnings (0.00 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
+------------------------------+
| greatest(int_type, str_type) |
+------------------------------+
| invalid_time                 |
+------------------------------+
1 row in set (0.00 sec)

+-----------------------------------+
| greatest(int_type, datetime_type) |
+-----------------------------------+
| 2020-01-01 00:00:00               |
+-----------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
+-----------------------------------+
| greatest(str_type, datetime_type) |
+-----------------------------------+
| 2020-01-01 00:00:00.000000        |
+-----------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
```

### 3. What did you see instead (Required)

In TiDB:

```sql
mysql> source /Users/jmpotato/test.sql
Query OK, 0 rows affected (0.03 sec)

Query OK, 0 rows affected (0.02 sec)

Query OK, 1 row affected (0.01 sec)

+----------+--------------+---------------------+
| int_type | str_type     | datetime_type       |
+----------+--------------+---------------------+
|    12345 | invalid_time | 2020-01-01 00:00:00 |
+----------+--------------+---------------------+
1 row in set (0.00 sec)

+------------------------------------------+
| least(int_type, str_type, datetime_type) |
+------------------------------------------+
|                                        0 |
+------------------------------------------+
1 row in set, 1 warning (0.00 sec)

+---------------------------+
| least(int_type, str_type) |
+---------------------------+
|                         0 |
+---------------------------+
1 row in set, 1 warning (0.00 sec)

+--------------------------------+
| least(int_type, datetime_type) |
+--------------------------------+
|                          12345 |
+--------------------------------+
1 row in set (0.00 sec)

+--------------------------------+
| least(str_type, datetime_type) |
+--------------------------------+
| invalid_time                   |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

+---------------------------------------------+
| greatest(int_type, str_type, datetime_type) |
+---------------------------------------------+
|                              20200101000000 |
+---------------------------------------------+
1 row in set, 1 warning (0.00 sec)

+------------------------------+
| greatest(int_type, str_type) |
+------------------------------+
|                        12345 |
+------------------------------+
1 row in set, 1 warning (0.01 sec)

+-----------------------------------+
| greatest(int_type, datetime_type) |
+-----------------------------------+
|                    20200101000000 |
+-----------------------------------+
1 row in set (0.00 sec)

+-----------------------------------+
| greatest(str_type, datetime_type) |
+-----------------------------------+
| 2020-01-01 00:00:00               |
+-----------------------------------+
1 row in set, 1 warning (0.00 sec)
```

### 4. Affected version (Required)

```
+ TiDB version:
Release Version: v4.0.0-beta.2-624-g4e74d4039-dirty
Edition: Community
Git Commit Hash: 4e74d40392e9b81f2a90cbc2f7777fd35f87b2e7
Git Branch: master
UTC Build Time: 2020-06-12 02:31:49
GoVersion: go1.14.3
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 900

## Mentor
 * @lzmhhh123

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: close #17994 

Problem Summary:

In MySQL, built-in function `least` and `greatest` use some different type conversion rules than others. You can check the differences from MySQL 8.0 Reference Manual:

* [Type Conversion in LEAST()](https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#function_least)
* [Type Conversion in Expression Evaluation](https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html)

In TiDB, these differences are not handled well so that some sql queries are incompatible with MySQL. You can see some cases from issue #17994 

### What is changed and how it works?

* The function least/greatest/between used to determine type conversion
* Related unit and integration tests

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note.

### CODE REVIEW

**Comment by JmPotato** [comment]
@sre-bot /run-all-tests

**Comment by JmPotato** [comment]
@sre-bot /run-all-tests

**Comment by JmPotato** [comment]
@sre-bot /run-check_dev_2

**Comment by JmPotato** [comment]
Because the type conversion rules changed for least and greatest...I'm wondering that some test cases in integration_common_test may not be correct any more. How should I get the test data in `github.com/pingcap/tidb-test` to compare with MySQL and update the test cases?

**Comment by JmPotato** [comment]
I want to know how the integration-common-test works? Does it compare query results with current master version TiDB or just with the fixed results written in it?

**Comment by JmPotato** [comment]
@sre-bot /run-check_dec_2

**Comment by JmPotato** [comment]
@sre-bot /run-check_dev_2

**Comment by JmPotato** [comment]
@sre-bot /run-check_dev

**Comment by JmPotato** [comment]
@XuHuaiyu @qw4990 PTAL

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests

File: expression/builtin_compare_test.go [hunk 1]
```diff
@@ -279,23 +279,27 @@ func (s *testEvaluatorSuite) TestGreatestLeastFuncs(c *C) {
...
+			curTimeString, "123", false, false,
+		},
+		{
+			[]interface{}{tm, tmWithFsp},
+			curTimeWithFspString, curTimeString, false, false,
 		},
 		{
-			[]interface{}{tm, "invalid_time_1", "invalid_time_2", tmWithFsp},
-			curTimeWithFspString, "invalid_time_1", false, false,
+			[]interface{}{tm, "invalid_time_a", "invalid_time_b", tmWithFsp},
```
Comment (ichn-hu) [inline expression/builtin_compare_test.go:297]: why change `invalid_time_1` to `..._a` ?

File: expression/builtin_compare_test.go [hunk 2]
```diff
@@ -279,23 +279,27 @@ func (s *testEvaluatorSuite) TestGreatestLeastFuncs(c *C) {
...
+			curTimeWithFspString, curTimeString, false, false,
 		},
 		{
-			[]interface{}{tm, "invalid_time_2", "invalid_time_1", tmWithFsp},
-			curTimeWithFspString, "invalid_time_2", false, false,
+			[]interface{}{"invalid_time_a", "invalid_time_b"},
+			"invalid_time_b", "invalid_time_a", false, false,
```
Comment (ichn-hu) [inline expression/builtin_compare_test.go:302]: this test case does not make any sense?

Comment (ichn-hu) [reply expression/builtin_compare_test.go:302]: Why not include your test cases in [this issue](https://github.com/pingcap/tidb/issues/17994)?

**Comment by ichn-hu** [comment]
> I want to know how the integration-common-test works? Does it compare query results with current master version TiDB or just with the fixed results written in it? If it's latter, maybe the test cases in it should be updated to be compatible with MySQL according to this pr.

this result is fixed, if the behavior change is intended, you might update the test suit.

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -382,8 +382,48 @@ func temporalWithDateAsNumEvalType(argTp *types.FieldType) (argEvalType types.Ev
...
+			isAllStr = false
+		}
+		cmpEvalType = getBaseCmpType(cmpEvalType, tp, lft, rft)
+		lft = rft
+	}
+	argTp = cmpEvalType
+	if isAllStr || cmpEvalType.IsStringKind() || (strFound && argTp == types.ETReal) {
+		argTp = types.ETString
+	}
+	if datetimeFound {
```
[the remark this answers is not in the document]
Comment (JmPotato) [reply expression/builtin_compare.go:418]: Sorry, it has been a while since I created this pr. I need to call my memory back...

**Comment by JmPotato** [comment]
> > I want to know how the integration-common-test works? Does it compare query results with current master version TiDB or just with the fixed results written in it? If it's latter, maybe the test cases in it should be updated to be compatible with MySQL according to this pr.
> 
> this result is fixed, if the behavior change is intended, you might update the test suit.

But as a contributor, it seems that I don't have the permission to read/write the test repo?

**Comment by ichn-hu** [comment]
That's right, you could create a PR in the test repo, and include the PR number when asking the CI bot to run test, an example is like this


https://github.com/pingcap/tidb/pull/13733#issuecomment-560275304








Best regards,


Zhifeng


&nbsp;




------------------&nbsp;Original&nbsp;------------------

**Comment by JmPotato** [comment]
> That's right, you could create a PR in the test repo, and include the PR number when asking the CI bot to run test, an example is like this

I mean I don't even have permission to visit https://github.com/pingcap/tidb-test/ in GitHub😢

**Comment by ichn-hu** [comment]
Well...That's a real problem...

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests

**Comment by JmPotato** [comment]
@sre-bot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@sre-bot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-common-test tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077 tikv=pr/7888 copr-test=pr/133

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077 tikv=pr/7888 copr-test=pr/133

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-all-tests tidb-test=pr/1077

**Comment by JmPotato** [comment]
@ti-srebot /run-integration-copr-test tikv=pr/7888 copr-test=pr/133

**Comment by JmPotato** [comment]
@ti-srebot /run-integration-copr-test

**Comment by fzhedu** [comment]
@nullnotnil please review this PR, because the comparison rules among different data types is conflusing in Mysql, tks.

**Comment by fzhedu** [comment]
can this PR solve https://github.com/pingcap/tidb/issues/19534 ?

File: expression/builtin_compare_vec.go [hunk 4]
```diff
@@ -737,14 +737,19 @@ func (b *builtinLeastTimeSig) vecEvalString(input *chunk.Chunk, result *chunk.Co
 			if dst.IsNull(i) {
 				continue
 			}
+			// If any invalid time is found, leastTime will use the string comprasion res
```
Comment (nullnotnil) [inline expression/builtin_compare_vec.go:740]:
```suggestion
			// If any invalid time is found, leastTime will use the string comparison res
```

**Comment by nullnotnil** [comment]
> @nullnotnil please review this PR, because the comparison rules among different data types is conflusing in Mysql, tks.

LGTM

**Pull request closed without merging**